mitodl/unified-ecommerce-frontend
# Add basket UI stuff (and a bunch of scaffolding)

_You will need the backend changes from https://github.com/mitodl/unified-ecommerce/pull/183 for this code to work._ 

### What are the relevant tickets?

https://github.com/mitodl/hq/issues/5656

### Description (What does it do?)

Adds two pages to the front end:
- A _basket chooser_ page, so we can throw you to something relevant if you just hit the ecommerce system.
- A _basket_ page, allowing you to view the things in the basket for a given system, apply discount codes, and check out.

This also adds a bunch of scaffolding stuff. I've pulled a handful of components from `ol-components` in Learn into this - these need to be moved to smoot-design at some point but _for the purposes of getting this to a state that people can see it_ they're just here for now. Similarly, the OpenAPI client is included while we work out the remaining issues with the pipeline build.

This also has a "force reauth" that allows you to force the app to re-do the `establish_session` call. This is for debugging only.

### Screenshots (if appropriate):

Desktop screenshots only:

Chooser page:
![image](https://github.com/user-attachments/assets/05c2d6b2-1dbb-44ab-b240-62172e264a5b)

Basket page:
![image](https://github.com/user-attachments/assets/09fa05e9-f56a-4f44-a1b3-bef148bff9f9)

Empty basket page:
![image](https://github.com/user-attachments/assets/b3741b7e-ede6-4b52-932e-1360dee26119)

### How can this be tested?

You'll need a working UE backend setup. Make sure you have data in it - run `generate_test_data` if you haven't already - and for best results you should also have some tax rates set up and some discounts. 

Build and run the app.
- In the chooser page, you should be able to choose a system and then see the basket that's present for it. If there isn't one, it should be created (this is an API thing but worth checking here too).
- In the basket page, you should be able to see:
   - An empty basket if there's nothing in it.
   - A basket showing a list of the items you have in the basket. 
      - You should also see a Order Summary card with a list of items in the cart (with any applied discounts), tax assessed (if any), and a total. 
      - You should be able to apply a discount via the Order Summary card - there is a field there, but it doesn't have any chrome - and if there's an error, you should see that. 
      - You should be able to click Place Order and it should take you to CyberSource.

To add items to a specific cart, use the test mule app: `http://ue.odl.local:9080/cart/<system slug>/` . You can also hit the API for this manually - `/api/v0/payments/baskets/create_from_product/{system_slug}/{sku}/` .

### Additional Context

The goal for this PR (and its related PR in `unified-ecommerce`) is to get something up and running so we can make sure the deployments work and so that people can more easily see progress in the system. Because of that, there's some shortcuts that were taken and some things aren't completely baked yet. Notably:

- There's some components that are located in this branch that really belong in smoot-design.
- The `Input` component is using the `InputBase` - this is already a WIP in smoot-design so opted to do a minimal amount here until it's ready.
- The API client is copied in rather than pulled from the automatically built version - we need to iron out a few kinks in that pipeline still. (If that happens before this PR is closed, we'll fix it in this PR.)
- The styling is at least sort of close but it's still pretty far off from what's in Figma. Some elements aren't described in Figma currently (the chooser page), and some data needs to be implemented still (product images), and the design as a whole isn't necessarily complete right now as it is.

Also, an important note for local deployment: this and the API need to live on the same(ish) hostnames. So, you should access this from `ue.odl.local:8072` and the API should be `ue.odl.local:9080`. (There _may_ be some references to the front end being at `uefe.odl.local` - this won't share an Origin with the API and you'll get CSRF and CORS errors.)
